### PR TITLE
🚧Create release workflow

### LINKED ISSUE #97: 🚧Create release workflow
### Complete The Item Below

- [X] I have updated the title without removing the 🚧 emoji.

### Description

Create a release workflow to run some checks, build, run tests, and then perform a GitHub release.

Make sure to have the release set to the _**latest**_ release.

Look [here](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release) at the REST API docs for information.

This should be done using the [GH Release (softprops)](https://github.com/marketplace/actions/gh-release) GitHub action from the marketplace.

After doing some digging, you cannot set releases to be the latest if it is **_draft_** or **_prerelease_**.  So when the action is in preview, the release to the marketplace will have to be manual.  Once we are v1.0, setting the releases are set to the latest by default and we will not have to manually set it to the marketplace.

### Acceptance Criteria

- [x] Release workflow created
    - [x] Only ran manually
    - [x] Checks to make sure that it is being run on the master branch only
    - [x] Pulls the version from the csproj file
    - [x] Use the version to make sure a tag that matches the version does not already exist
    - [x] Check that the release notes for the given version exist
    - [x] Run build
    - [x] Run tests
    - [x] Perform GitHub release using GitHub action
    - [x] Validate tag and release notes checks use infrastructure scripts
- [x] Validate tag and release notes checks use infrastructure scripts for _**release.yml**_ workflow

### ToDo Items

- [X] Change type labels added to this issue.  Refer to the _**Change Type Labels**_ section below.
- [X] Priority label added to this issue.  Refer to the _**Priority Type Labels**_ section below.
- [X] Issue linked to the correct project _(if applicable)_.
- [X] Issue linked to the correct milestone _(if applicable)_.
- [x] Draft pull request created and linked to this issue _(only required with code changes)_.

### Issue Dependencies

_No response_

### Related Work

_No response_

### Additional Information:

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type         | Label                |
|---------------------|----------------------|
| Bug Fixes           | `🐛bug` |
| Breaking Changes    | `🧨breaking changes` |
| New Feature         | `✨new feature` |
| CICD Changes        | `♻️cicd` |
| Code Doc Changes    | `🗒️documentation/code` |
| Product Doc Changes | `📝documentation/product` |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label             |
|---------------------|-------------------|
| Low Priority        | `low priority`    |
| Medium Priority     | `medium priority` |
| High Priority       | `high priority`   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### PR DESCRIPTION
**Pull request by kselena**
<h1 style="font-weight:bold" align="center">Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: feature/<issue-num>-<description>
      Example: feature/123-my-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_development_** branch with the name **_develop_**.

- [x] The **_base(target)_** branch for this pull request is a **_development_** branch with the name **_develop_**.
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

💡For more information on branching, refer to the project's [branching documentation](https://github.com/KinsonDigital/.github/blob/master/docs/BranchingFullGitFlow.md).

- [x] The associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] The pull request title matches the title of the associated issue.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] The assigned project matches the project of the associated issue.

- [x] The assigned milestone matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_develop_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

**Comment by kselena** [comment]
> **Note** PR and issue put on https://github.com/KinsonDigital/BranchValidator/labels/on%20hold
> Reason: Waiting for the cicd scripts to be added to the org repo and to update the release workflow

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        repo-name: BranchValidator
+        repo-token: ${{ secrets.CICD_TOKEN }}
+        branch-name: "${{ github.ref }}"
+        trim-start-from-branch: "refs/heads/"
+        file-format: xml # Not case sensitive
+        file-path: "BranchValidator/BranchValidator.csproj"
+        version-keys: Version
+
+    - name: Validate Tag
+      run: deno run --allow-run "${{ github.workspace }}/.github/cicd/validate-tag.ts" "${{ inputs.release_type }}" "v${{ steps.get-version.outputs.version }}";
```
Comment (KinsonDigitalAdmin) [inline .github/workflows/release.yml:63]: Update the script path to a URL to the script in the infrastructure repository.
This is currently on hold to verify that it works with package monster.  

> **Warning** Do not update this until we have proven the concept with PackageMonster first.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    - name: Validate Release Notes
+      run: deno run --allow-read "${{ github.workspace }}/.github/cicd/validate-notes.ts" "${{ inputs.release_type }}" "v${{ steps.get-version.outputs.version }}";
```
Comment (KinsonDigitalAdmin) [inline .github/workflows/release.yml:66]: Update the script path to a URL to the script in the infrastructure repository.
This is currently on hold to verify that it works with package monster.  

> **Warning** Do not update this until we have proven the concept with PackageMonster first.

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]
Update the release workflow to use infrastructure scripts.
Tasks added to issue.

> **Warning** Do not do this until we have proven the concept with PackageMonster and the issue has been lifted out of https://github.com/KinsonDigital/BranchValidator/labels/on%20hold status

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    runs-on: ubuntu-latest
+    steps:
+    - name: Validate Branch
+      run: |
+        $branch = "${{ github.ref }}";
+
+        if ($branch -ne "refs/heads/master") {
+          Write-Host "::error::Release workflow can only run on the $branch branch";
+          exit 1;
+        }
```
[the remark this answers is not in the document]
Comment (KinsonDigitalAdmin) [reply .github/workflows/release.yml:33]: This change was not done but the 👍🏻 emoji is there.  
Did you forget to push up the changes?

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Comment by KinsonDigitalAdmin** [comment]
> **Note** Taken off of https://github.com/KinsonDigital/BranchValidator/labels/on%20hold status.  Concept with org level deno scripts works perfectly.

File: .github/workflows/release.yml (hunk 2 above)
Comment (KinsonDigitalAdmin) [inline .github/workflows/release.yml:66]: 1. Update the script paths for both steps to a URL to the script in the infrastructure repository.
This is currently on hold to verify that it works with package monster.  
Tasks added to issue.
2. Update to multi-line shell scripts

> **Warning** Do not update this until we have proven the concept with PackageMonster first and the issue has been lifted out of https://github.com/KinsonDigital/BranchValidator/labels/on%20hold status

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]
